h2o-mapreduce-generic/src/main/java/water/hadoop/EmbeddedH2OConfig.java

```diff
@@ -0,0 +1,170 @@
+    return true;
+  }
+
+  @Override
+  public String fetchFlatfile() throws Exception {
+    System.out.println("EmbeddedH2OConfig: fetchFlatfile called");
+    MapperToDriverMessage msg = new MapperToDriverMessage();
+    msg.setMessageFetchFlatfile(_embeddedWebServerIp, _embeddedWebServerPort);
+    Socket s = new Socket(_driverCallbackIp, _driverCallbackPort);
+    msg.write(s);
+    DriverToMapperMessage msg2 = new DriverToMapperMessage();
+    msg2.read(s);
+    char type = msg2.getType();
+    if (type != DriverToMapperMessage.TYPE_FETCH_FLATFILE_RESPONSE) {
+      int typeAsInt = (int)type & 0xff;
+      String str = "DriverToMapperMessage type unrecognized (" + typeAsInt + ")";
+      Log.err(str);
+      throw new Exception(str);
+    }
+    s.close();
+    String flatfile = msg2.getFlatfile();
+    System.out.println("EmbeddedH2OConfig: fetchFlatfile returned");
+    System.out.println("------------------------------------------------------------");
+    System.out.println(flatfile);
+    System.out.println("------------------------------------------------------------");
+    return flatfile;
+  }
+
+  @Override
+  public void notifyAboutCloudSize(InetAddress ip, int port, InetAddress leaderIp, int leaderPort, int size) {
+    _embeddedWebServerIp = ip.getHostAddress();
+    _embeddedWebServerPort = port;
+
+    try {
+      MapperToDriverMessage msg = new MapperToDriverMessage();
+      msg.setDriverCallbackIpPort(_driverCallbackIp, _driverCallbackPort);
+      msg.setMessageCloudSize(ip.getHostAddress(), port, leaderIp.getHostAddress(), leaderPort, size);
+      BackgroundWriterThread bwt = new BackgroundWriterThread();
+      System.out.printf("EmbeddedH2OConfig: notifyAboutCloudSize called (%s, %d, %d)\n", ip.getHostAddress(), port, size);
+      bwt.setMessage(msg);
+      bwt.start();
+    }
+    catch (Exception e) {
+      System.out.println("EmbeddedH2OConfig: notifyAboutCloudSize caught an Exception");
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void exit(int status) {
+    try {
+      MapperToDriverMessage msg = new MapperToDriverMessage();
+      msg.setDriverCallbackIpPort(_driverCallbackIp, _driverCallbackPort);
+      msg.setMessageExit(_embeddedWebServerIp, _embeddedWebServerPort, status);
+      System.out.printf("EmbeddedH2OConfig: exit called (%d)\n", status);
+      BackgroundWriterThread bwt = new BackgroundWriterThread();
+      bwt.setMessage(msg);
+      bwt.start();
+      System.out.println("EmbeddedH2OConfig: after bwt.start()");
+    }
+    catch (Exception e) {
+      System.out.println("EmbeddedH2OConfig: exit caught an exception 1");
+      e.printStackTrace();
+    }
+
+    try {
+      // Wait one second to deliver the message before exiting.
+      Thread.sleep (1000);
+      Socket s = new Socket("127.0.0.1", _mapperCallbackPort);
+      byte[] b = new byte[] { (byte) status };
+      OutputStream os = s.getOutputStream();
+      os.write(b);
+      os.flush();
+      s.close();
```

Comment:
Minor: resources not safely closed. Use try-with-resources or close them in finally block would be better.